internal/controllers/state/statusupdater.go

```diff
@@ -31,40 +53,71 @@ func NewStatusUpdater(kubeClient clients.ClientsInterface) StatusUpdater {
 	}
 }
 
-// UpdateWithState updates sr's Status.State property with state, and updates the object in Kubernetes.
-// TODO(qbarrand) make this function return an error
-func (su *statusUpdater) UpdateWithState(ctx context.Context, sr *v1beta1.SpecialResource, state string) {
+// SetAsProgressing changes SpecialResource's Progressing condition as true and changes Ready and Errored conditions to false, and updates the status in the API.
+func (su *statusUpdater) SetAsProgressing(ctx context.Context, sr *v1beta1.SpecialResource, reason, message string) error {
+	return su.updateWithMutator(ctx, sr, func(o *v1beta1.SpecialResource) {
```

Comment:
Why have a `updateWithMutator`? Could we just set the status and then use a [`StatusWriter`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#StatusWriter) to save the SpecialResource in Kubernetes?

Reply:
I'll investigate it, thanks for the idea

Reply:
Please take a look